workflow/controller/operator.go

```diff
@@ -247,6 +258,49 @@ func (woc *wfOperationCtx) persistUpdates() {
 	}
 }
 
+// processNodesWithRetries checks whether any of the failed nodes had retries left and if so,
+// starts a new pod. It returns whether there are retries left or not.
+func (woc *wfOperationCtx) processNodesWithRetries() (bool, error) {
+	// TODO(shri): This should be fixed by moving the container execution logic into a
+	// separate layer. Steps, Retries, etc. will be higher layers that call into the
+	// container execution layer.
+	for _, node := range woc.wf.Status.GetNodesWithRetries() {
+		if node.Completed() {
+			continue
+		}
+		lastChildNode, err := woc.getLastChildNode(&node)
+		if err != nil {
+			return false, fmt.Errorf("Failed to find last child of node " + node.Name)
+		}
+
+		if !lastChildNode.Completed() {
+			// last child node is still running.
+			return false, nil
+		}
+
+		if lastChildNode.Successful() {
+			woc.markNodePhase(node.Name, wfv1.NodeSucceeded)
+			return false, nil
+		}
+
+		if !lastChildNode.CanRetry() {
+			woc.log.Infof("Node cannot be retried. Marking it failed")
+			woc.markNodePhase(node.Name, wfv1.NodeFailed)
+			return false, nil
+		}
+
+		if int32(len(node.Children)) > node.RetryInfo.Limit {
+			woc.log.Infoln("No more retries left. Failing...")
+			woc.markNodePhase(node.Name, wfv1.NodeFailed)
```

Comment:
Please propagate the failure reason when marking nodes as failed.

Reply:
Done.